### ISSUE TITLE
Nodes get wrong sortorder

### ISSUE THREAD

**Issue by joepvtl** [umbraco/Umbraco-CMS#8359]
When you first create a media item then a new content item, the content item gets the wrong sort order. It gets fixed after an AppPool recycle. We made this conclusion because we added the SqlProfiler when creating a new content node and saw that it looked at the NodeObjectId of an media item to determine the SortOrder, but it did create the content correctly.
There aren't any error logs before this bug occurs.

## Umbraco version

I have tested this on 8.6.2

Reproduction
------------

### Steps to reproduce

Content gets wrong SortOrder:
1. Setup a fresh install of umbraco and use the starter kit.
2. Create a new media item.
3. Create a new conten item under the 'About us' node.

Media gets wrong SortOrder:
1. Setup a fresh install of umbraco and use the starter kit.
2. Create a new content item.
3. Create a new media item under the 'People' node.

After the last step you can see that the node doesn't get placed at the bottom where it should be. When you look in the database you can see that the node got a SortOrder 0, instead of the expected 2.

### Expected result

Nodes get the correct sortorder.

### Actual result

Nodes get the wrong sortorder.

---
_This item has been added to our backlog [AB#7545](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/7545)_

**Comment by nul800sebastiaan** [comment]
Thanks @joepvtl - I fail to see how this could possibly get fixed after the app pool recycles? It's not like we go in the database after an app pool recycle and update the sortorder column.. can you clarify?

**Comment by joepvtl** [comment]
Well, I think it has something to do with SqlTemplates. Since they both use the same template key to determine the SortOrder and the template gets put into a dictionary. When the content gets created first the query for determining the SortOrder of content gets put into the dictionary, then the media uses that same key and basically uses the same query.

**Comment by alexroode** [comment]
We are seeing a similar issue. In our case we have a news feed node configured as a List View whose child posts are sorted by their sort order descending, to show the newest children first while still allowing editors to manually sort if needed.  Sometimes our site seems to get into a state where each new child item gets created with sort order 0 instead of incrementing like we expect. In this case editors are forced to manually sort to make them appear before the older children with nonzero sort order, so this causes a problem for us.

If we restart the app pool, then create a new node, it seems to properly increment the next new node's sort order. But we seem to eventually get back to the state described above - I have not been able to pinpoint when this happens.

**Comment by joepvtl** [comment]
@alexroode have you tried the reproduction steps on your site?

**Comment by nul800sebastiaan** [comment]
```
Content gets wrong SortOrder:
- Setup a fresh install of umbraco and use the starter kit.
- Create a new media item.
- Create a new conten item under the 'About us' node.
```

I can not reproduce this in v 8.6.3, here's what is being stored, `test1`, `test2` and `test3` are content items, between each content item I created a media item:

![image](https://user-images.githubusercontent.com/304656/88154731-84055200-cc07-11ea-8268-d17b66c75ba6.png)

```
Media gets wrong SortOrder:
- Setup a fresh install of umbraco and use the starter kit.
- Create a new content item.
- Create a new media item under the 'People' node.
```

As you can already see in the screenshot above, all media items seem to get the sortorder of 1, which is what I get when uploading through the media section, that must be a mistake somewhere!

The problem persists when I create media items from a media picker in the content section. 

Aha.. I created a folder in the media section, recycled the app pool and tried again. After that the sort order for media was suddenly correct, then I created a new content item and indeed the sort order went to 0. Yikes. 

Top results = content, bottom results = media

![image](https://user-images.githubusercontent.com/304656/88156092-3853a800-cc09-11ea-9d60-94ab6d3a8301.png)

We'll have to have a look at what's happening here, thanks for the report @joepvtl!

**Comment by ronaldbarendse** [comment]
I can confirm this issue, but it's hit and miss, as some children have the right sort order, but others all have 0. All content nodes are culture variant, but as it's also happening to media, that shouldn't make a difference.

![List view](https://user-images.githubusercontent.com/1051287/94238701-06183c80-ff11-11ea-8a1a-ba5861d0012f.png)

@joepvtl is probably right about the SqlTemplates, as the `first` parameter, but also the `NodeObjectTypeId` property (that differs per repository type) are not SQL parameters/arguments and get included in the generic `Umbraco.Core.VersionableRepository.GetSortOrder` template:
https://github.com/umbraco/Umbraco-CMS/blob/d65041c2503fc85ac4619a4abcd703dcc20f067d/src/Umbraco.Core/Persistence/Repositories/Implement/ContentRepositoryBase.cs#L901-L908

This also makes total sense for the application restart to fix things, as the `NodeObjectTypeId` of the first repository doing a `GetNewChildSortOrder()` gets used for all next queries!

**Comment by joepvtl** [comment]
Hey @nul800sebastiaan, do you have any update regarding this issue? I see it has been added to the backlog?

**Comment by nul800sebastiaan** [comment]
This is fixed for 8.11.0 in PR https://github.com/umbraco/Umbraco-CMS/pull/8961
